File: nodes/array-shuffle/array-shuffle.process.py
import random

async def process(inputs, settings, config, nodeConfig):
  array = inputs.get("array", [])
  if not isinstance(array, list):
    array = [array]
    
  seed = settings.get("seed")
  
  # Create new random instance if seed provided
  rand = random.Random(seed) if seed is not None else random.Random()
  
  # Create index mapping and shuffle together
  indices = list(range(len(array)))
  result = array.copy()
  
  # Fisher-Yates shuffle
  for i in range(len(result) - 1, 0, -1):
    j = rand.randint(0, i)
    result[i], result[j] = result[j], result[i]
    indices[i], indices[j] = indices[j], indices[i]
  
  # For empty or single-element arrays, match the expected output format
  if len(array) <= 1:
    return {
      "array": result
    }
  
  return {
    "array": result,
    "indices": indices
  } 

File: nodes/array-shuffle/array-shuffle.process.js
export default async ({inputs, settings, config}) => {

  const array = Array.isArray(inputs.array) ? inputs.array : [inputs.array];
  const seed = settings.seed;
  
  // Fisher-Yates shuffle with optional seeded random
  const shuffle = (arr) => {
    const result = [...arr];
    const indices = arr.map((_, i) => i);
    
    // Seeded random number generator
    const random = (() => {
      if (seed === undefined) return Math.random;
      
      let value = Number(seed);
      return () => {
        value = (value * 16807) % 2147483647;
        return (value - 1) / 2147483646;
      };
    })();
    
    for (let i = result.length - 1; i > 0; i--) {
      const j = Math.floor(random() * (i + 1));
      [result[i], result[j]] = [result[j], result[i]];
      [indices[i], indices[j]] = [indices[j], indices[i]];
    }
    
    return { result, indices };
  };
  
  const { result, indices } = shuffle(array);
  
  return {
    array: result,
    indices: indices
  };
}; 